android_unpinner/__main__.py

```diff
@@ -102,6 +104,71 @@ def install_apk(apk_files: list[Path]) -> None:
         adb(f"install --no-incremental {apk_files[0]}")
 
 
+def find_apks_in_xapk(xapk_path: Path, output_dir = None) -> list[Path] | None:
+    """
+    Extracts APK files from an XAPK file to a folder and returns their paths.
+    """
+
+    if not xapk_path.name.lower().endswith(".xapk"):
+        return None
+
+    if not os.path.exists(xapk_path):
+        return None
+    
+    logging.info(f"Processing XAPK: {os.path.basename(xapk_path)}")
+    
+    if output_dir is None:
+        base_name = os.path.basename(xapk_path)
+        dir_name = "XAPKs" + os.path.sep + os.path.splitext(base_name)[0].replace(" ", "_") + "_extracted"
+        extraction_dir = os.path.join(os.path.dirname(xapk_path), dir_name)
+    else:
+        extraction_dir = os.path.abspath(output_dir)
+
+    if os.path.exists(extraction_dir):
+        logging.warning(f"Directory '{extraction_dir}' already exists. New files will be merged/overwritten.")
+    else:
+        os.makedirs(extraction_dir)
+        logging.info(f"Created extraction directory: {extraction_dir}")
+
+    apk_files = []
+
+    try:        
+        with zipfile.ZipFile(xapk_path, 'r') as zip_ref:
+            zip_ref.extractall(extraction_dir)
+            logging.info("XAPK extraction complete.")
+
+        logging.info("Searching for APK files...")
+        for root, _, files in os.walk(extraction_dir):
+            for file_name in files:
+                if file_name.lower().endswith(".apk"):
+                    full_path = os.path.join(root, file_name)
+                    apk_files.append(Path(full_path))
+                    logging.info(f"Found APK: {full_path}")
+                    
+        return apk_files
+
+    except zipfile.BadZipFile:
+        logging.error(f"Error: '{xapk_path}' is not a valid ZIP file.")
+        return None
+    except Exception as e:
+        logging.error(f"An unexpected error occurred: {e}")
+        return None
+
+
+def process_xapks(apk_files: list[Path]) -> list[Path]:
+    """
+    Preprocess the list of APK files to handle any XAPK files.
+    """
+    apk_files = list(apk_files)
+    for apk in apk_files:
+        if isinstance(apk, Path) and apk.suffix.lower() == ".xapk":
+            apk_files.remove(apk)
```

Comment:
Modifying a list while iterating over it is a can of worms. Add a new `ret = []` list and populate that.